### PR TITLE
User rest tests

### PR DESCRIPTION
**Pull request by cklanac**
Updated:
- Added validation to UserSchema and remove some validation from routes
- Moved bcrypt hash to UserSchema.pre so it run on all .save() calls, including tests. This eliminates the issues of adding bcrypt to the test file, or calling 'post' to add dummy users to the tests
- Added test-main.js, pretty standard stuff
- Implemented get, put and delete routes
- Fixed Basic Strategy so it doesn't fire second .then() on username not found failure
- Added autorun for mocha in package.json `"test": "nodemon -x mocha"`
- Created lots of tests
- Minor updates to config.js

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -24,8 +26,9 @@ app.use('*', function(req, res) {
 // assumes runServer has run and set `server` to a server object
 let server;
 
-function runServer() {
+function runServer(callback) {
```
Comment (benjaminEwhite) [inline server.js]: Why add this param?

File: test/test-users.js [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+const spies = require('chai-spies');
+const mongoose = require('mongoose');
+const app = require('../server').app;
+
+const {
+  User
+} = require('../users/models');
+const should = chai.should();
+
+var UrlPattern = require('url-pattern');
```
Comment (benjaminEwhite) [inline test/test-users.js]: `const`

File: test/test-users.js [hunk 3]
```diff
@@ -0,0 +1,402 @@
+const chai = require('chai');
+const chaiHttp = require('chai-http');
+const spies = require('chai-spies');
+const mongoose = require('mongoose');
+const app = require('../server').app;
```
Comment (benjaminEwhite) [inline test/test-users.js]: i like to try to keep 3rd party imports together in a group, separated by blank line, then imports of app specific stuff, both alphabetical.

File: test/test-users.js [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+  "/users/:userId": new UrlPattern('/users/:userId')
+}
+
+chai.use(chaiHttp);
+chai.use(spies);
+
+describe('Test /user endpoints with empty database', () => {
+  beforeEach((done) => {
+    // Approach #1 Drop database and `ensureIndexes()`
+    mongoose.connection.db.dropDatabase()
```
Comment (benjaminEwhite) [inline test/test-users.js]: We use `mongoose.connection.dropDatabase();` over `mongoose.connection.db.dropDatabase()` elsewhere. Think `db` is already implied in connection, so let's stick with `mongoose.connection.dropDatabase()` here, unless you have a strong objection.

File: test/test-users.js [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+const pattern = {
+  "/users": new UrlPattern('/users'),
+  "/users/:userId": new UrlPattern('/users/:userId')
+}
+
+chai.use(chaiHttp);
+chai.use(spies);
+
+describe('Test /user endpoints with empty database', () => {
+  beforeEach((done) => {
```
Comment (benjaminEwhite) [inline test/test-users.js]: We're consistently  using approach of returning a promise, not passing/calling a `done` callback in other apps and would like to retain that here.

File: test/test-users.js [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+    // Approach #1 Drop database and `ensureIndexes()`
+    mongoose.connection.db.dropDatabase()
+      .then(() => User.ensureIndexes())
+      .then(() => done());
```
Comment (benjaminEwhite) [inline test/test-users.js]: return promise instead of call `done`.

File: test/test-users.js [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+
+    // Approach #2 Remove items, retains indexes
+    // User.remove({}).then( () => done())
```
Comment (benjaminEwhite) [inline test/test-users.js]: ditto

File: test/test-users.js [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+
+  });
+  describe('GET /users', () => {
+    it('should return an empty list of users initially', () => {
```
Comment (benjaminEwhite) [inline test/test-users.js]: nitpicking, but think `should return an empty list` is explicit enough, given parent `describe`.

File: test/test-users.js [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+          res.body.should.be.an('array');
+          res.body.length.should.equal(0);
+        });
+    });
+  });
+});
+
+describe('Test /user endpoints with prepopulated users', () => {
+  // config dummy users to be used throughout across the tests
+  let johnDoe = {
```
Comment (benjaminEwhite) [inline test/test-users.js]: might as well be `const` since never reassigned.

File: test/test-users.js [hunk 10]
```diff
@@ -0,0 +1,402 @@
...
+    username: 'john.doe@example.com',
+    password: 'letmein',
+    firstName: 'John',
+    lastName: 'Doe'
+  };
+  let janeDoe = {
```
Comment (benjaminEwhite) [inline test/test-users.js]: ditto `const`

File: test/test-users.js [hunk 11]
```diff
@@ -0,0 +1,402 @@
...
+    username: 'jane.doe@example.com',
+    password: 'password',
+    firstName: 'Jane',
+    lastName: 'Doe'
+  };
+
+  beforeEach((done) => {
```
Comment (benjaminEwhite) [inline test/test-users.js]: ditto return promise, not callback.

File: test/test-users.js [hunk 12]
```diff
@@ -0,0 +1,402 @@
...
+    // Approach #2.5 Remove items and then insert dummy users
+    User.remove({})
```
Comment (benjaminEwhite) [inline test/test-users.js]: `return`

File: test/test-users.js [hunk 13]
```diff
@@ -0,0 +1,402 @@
...
+      .then(() => User.create([johnDoe, janeDoe]))
+      .then(() => done())
```
Comment (benjaminEwhite) [inline test/test-users.js]: ditto

File: test/test-users.js (hunk 5 above)
Comment (benjaminEwhite) [reply test/test-users.js]: Would like to maintain this sequence from earlier in course:

```
before(function() {
    return runServer(TEST_DATABASE_URL);
  });

  beforeEach(function() {
    return promiseThatSeedsDb();
  });

  afterEach(function() {
    return tearDownDb();
  });

  after(function() {
    return closeServer();
  })
```

Crucially, we're tearing down database between each test. There should be zero interdependencies between tests.

File: test/test-users.js [hunk 14]
```diff
@@ -0,0 +1,402 @@
...
+    // Approach #1 Drop database and `ensureIndexes()`
+    mongoose.connection.db.dropDatabase()
+      .then(() => User.ensureIndexes())
+      .then(() => done());
+
+    // Approach #2 Remove items, retains indexes
+    // User.remove({}).then( () => done())
+
+  });
+  describe('GET /users', () => {
```
Comment (benjaminEwhite) [inline test/test-users.js]: Why nest here? since there's only one test in this suite, seems like no need to have nested describe  blocks.

File: test/test-users.js (hunk 12 above)
Comment (benjaminEwhite) [inline test/test-users.js]: Should never need to remove from db in a `beforeEach` step. Should always assume db has been wiped.

File: test/test-users.js [hunk 15]
```diff
@@ -0,0 +1,402 @@
...
+var UrlPattern = require('url-pattern');
+const pattern = {
+  "/users": new UrlPattern('/users'),
+  "/users/:userId": new UrlPattern('/users/:userId')
+}
+
+chai.use(chaiHttp);
+chai.use(spies);
+
+describe('Test /user endpoints with empty database', () => {
```
Comment (benjaminEwhite) [inline test/test-users.js]: Why even test this scenario? Kind of edge casey -- there's normally gonna be users in the db. Doesn't really test business logic so much.

File: test/test-users.js [hunk 16]
```diff
@@ -0,0 +1,402 @@
...
+        return chai.request(app).get('/users')
+          .then(res => {
+            res.should.have.status(200);
+            res.type.should.equal('application/json');
+            res.charset.should.equal('utf-8');
+
+            res.body.should.be.an('array');
+            res.body.length.should.equal(2); //there are 2 dummy users in our DB
+
+            res.body[0].should.be.an('object');
```
Comment (benjaminEwhite) [inline test/test-users.js]: might as well do `res.body.forEach((item) => {...})`

File: test/test-users.js [hunk 17]
```diff
@@ -0,0 +1,402 @@
...
+
+            // no password returned
+            res.body[0].should.not.have.property('password');
+          });
+      });
+    });
+
+    describe('POST', () => {
+      it('should allow adding a user', () => {
+        let newUser = {
```
Comment (benjaminEwhite) [inline test/test-users.js]: `const`

File: test/test-users.js [hunk 18]
```diff
@@ -0,0 +1,402 @@
...
+            res.should.have.status(201);
+            res.type.should.equal('application/json');
+            res.charset.should.equal('utf-8');
+
+            res.should.have.header('location');
+
+            res.body.should.be.an('object');
+            res.body.should.be.empty;
+
+            let params = pattern["/users/:userId"].match(res.headers.location);
```
Comment (benjaminEwhite) [inline test/test-users.js]: `const`

File: test/test-users.js [hunk 19]
```diff
@@ -0,0 +1,402 @@
...
+          })
+          .then(result => {
+            // Check that the user exists in the database
+            should.exist(result);
+            result.should.have.property('username');
+            result.username.should.equal(newUser.username);
+          });
+      });
+      it('should reject users without a username', () => {
+        let badUser = {
```
Comment (benjaminEwhite) [inline test/test-users.js]: `const`

File: test/test-users.js [hunk 20]
```diff
@@ -0,0 +1,402 @@
...
+          password: 'password',
+          firstName: 'bad',
+          lastName: 'user',
+        };
+        let spy = chai.spy();
```
Comment (benjaminEwhite) [inline test/test-users.js]: can this be a const?

File: test/test-users.js [hunk 21]
```diff
@@ -0,0 +1,402 @@
...
+        // Add a user without a username
+        return chai.request(app).post('/users').send(badUser)
+          .then(spy)
```
Comment (benjaminEwhite) [inline test/test-users.js:134]: Why do we need a spy here?

File: test/test-users.js [hunk 22]
```diff
@@ -0,0 +1,402 @@
...
+          .catch(err => {
+            // If the request fails, make sure it contains the error
+            let res = err.response;
+            res.should.have.status(422);
+            res.type.should.equal('application/json');
+            res.charset.should.equal('utf-8');
+            res.body.should.be.an('object');
+            res.body.should.have.property('message');
+            res.body.name.should.equal('ValidationError');
+            res.body.message.should.equal('User validation failed');
```
Comment (benjaminEwhite) [inline test/test-users.js]: Let's get rid of this assertion. Don't want a test that fails just because we update the response message.

File: users/models.js [hunk 23]
```diff
@@ -9,11 +9,16 @@ const UserSchema = mongoose.Schema({
...
+    // match: /[a-zA-Z0-9]/, // name must be alpha characters
+    match: /^([\w-\.]+@([\w-]+\.)+[\w-]{2,4})?$/, // validate email address
+    index: { unique: true }
   },
   password: {
-    type: String,
-    required: true
+    type: String, // besides setting type, it also protects against injection attacks http://stackoverflow.com/a/38415968
+    required: true,
+    trim: true,
```
Comment (benjaminEwhite) [inline users/models.js:18]: cool good to know

File: users/models.js [hunk 24]
```diff
@@ -9,11 +9,16 @@ const UserSchema = mongoose.Schema({
...
+    validate: [(pwd) => 6 < pwd.length, "Password must be at least 6 characters"]
```
Comment (benjaminEwhite) [inline users/models.js]: also good to know

File: users/models.js [hunk 25]
```diff
@@ -33,12 +38,17 @@ UserSchema.methods.validatePassword = function(password) {
     .then(isValid => isValid);
 }
 
-UserSchema.statics.hashPassword = function(password) {
-  return bcrypt
-    .hash(password, 10)
-    .then(hash => hash);
-}
+UserSchema.pre('save', function(next) {
+    var user = this;
```
Comment (benjaminEwhite) [inline users/models.js]: `const user = this;`

File: users/models.js [hunk 26]
```diff
@@ -33,12 +38,17 @@ UserSchema.methods.validatePassword = function(password) {
...
+    // only hash the password if it has been modified (or is new)
+    if (!user.isModified('password')) return next();
+    bcrypt.hash(user.password, 10, function(err, hash) {
+      if (err) return next(err);
+      user.password = hash;
```
Comment (benjaminEwhite) [inline users/models.js]: Is the idea that `user` sticks around? Trying to grok who cares about `user.password` since we don't return it or do anything with it.

File: users/router.js [hunk 27]
```diff
@@ -9,90 +9,46 @@ const router = express.Router();
...
     .then(user => {
-      return res.status(201).json(user.apiRepr());
+      return res.location('/users/' + user._id).status(201).json({});
     })
     .catch(err => {
-      res.status(500).json({message: 'Internal server error'})
+      // console.dir(err, {colors: true})
+      if (err.name === 'ValidationError') {
+          return res.status(422).json(err)
+      } else if (err.name === 'MongoError') {
```
Comment (benjaminEwhite) [inline users/router.js:48]: Do we really want to be leaking Mongo errors to the outside world?

File: users/router.js [hunk 28]
```diff
@@ -101,45 +57,74 @@ router.post('/', (req, res) => {
...
-      if (!isValid) {
-        return callback(null, false, {message: 'Incorrect password'});
+        return callback(null, false);
       }
-      else {
-        return callback(null, user)
-      }
-    });
+
+      user.validatePassword(password)
```
Comment (benjaminEwhite) [inline users/router.js]: Prefer to do

```
  return user.validatePassword(password);
)
.then(isValid => { })
```

as before just to avoid a layer of nesting

File: users/router.js [hunk 29]
```diff
@@ -101,45 +57,74 @@ router.post('/', (req, res) => {
...
+
+    // document.save triggers mongoose middleware such as model.pre
+    req.user.save()
+    .then(user => {
+      return res.json(user.apiRepr());
+    })
+    .catch(err => {
+      if (err.name === 'ValidationError') {
+          return res.status(422).json(err)
+      } else if (err.name === 'MongoError') {
```
Comment (benjaminEwhite) [inline users/router.js]: Ditto re: leaking MongoErrors

File: users/router.js [hunk 30]
```diff
@@ -101,45 +57,74 @@ router.post('/', (req, res) => {
...
 );
 
+router.delete('/me', passport.authenticate('basic', {session: false }),
+  (req, res) => {
+    req.user.remove()
+    .then(user => {
+      return res.json({});
+    })
+    .catch(err => {
+      res.status(500).json({message: 'Internal server error'})
```
Comment (benjaminEwhite) [inline users/router.js]: `res.status(500).json({message: 'Internal server error'});`

File: users/router.js [hunk 31]
```diff
@@ -101,45 +57,74 @@ router.post('/', (req, res) => {
...
+      } else {
+        return res.status(500).json({message: 'Internal server error'})
+      }
+    });
+  }
 );
 
+router.delete('/me', passport.authenticate('basic', {session: false }),
+  (req, res) => {
+    req.user.remove()
```
Comment (benjaminEwhite) [inline users/router.js:112]: semi-colon at end

File: users/router.js [hunk 32]
```diff
@@ -101,45 +57,74 @@ router.post('/', (req, res) => {
...
+    .then(user => {
+      return res.json(user.apiRepr());
+    })
+    .catch(err => {
+      if (err.name === 'ValidationError') {
+          return res.status(422).json(err)
+      } else if (err.name === 'MongoError') {
+          return res.status(422).json(err)
+      } else {
+        return res.status(500).json({message: 'Internal server error'})
```
Comment (benjaminEwhite) [inline users/router.js]: semi-colon

File: users/router.js [hunk 33]
```diff
@@ -101,45 +57,74 @@ router.post('/', (req, res) => {
...
+    // document.save triggers mongoose middleware such as model.pre
+    req.user.save()
+    .then(user => {
+      return res.json(user.apiRepr());
+    })
+    .catch(err => {
+      if (err.name === 'ValidationError') {
+          return res.status(422).json(err)
+      } else if (err.name === 'MongoError') {
+          return res.status(422).json(err)
```
Comment (benjaminEwhite) [inline users/router.js]: semi-colon

File: users/router.js [hunk 34]
```diff
@@ -101,45 +57,74 @@ router.post('/', (req, res) => {
...
+    req.user = Object.assign(req.user, req.body);
+
+    // document.save triggers mongoose middleware such as model.pre
+    req.user.save()
+    .then(user => {
+      return res.json(user.apiRepr());
+    })
+    .catch(err => {
+      if (err.name === 'ValidationError') {
+          return res.status(422).json(err)
```
Comment (benjaminEwhite) [inline users/router.js]: semi-colon

File: test/test-main.js [hunk 35]
```diff
@@ -0,0 +1,10 @@
+global.DATABASE_URL = 'mongodb://localhost/basic-auth-test'
+global.PORT = '8081'
+
+var runServer = require('../server').runServer;
```
Comment (benjaminEwhite) [inline test/test-main.js]: keep with style elsewhere in curriculum:

```javascript
const {runServer} = require('../server');
```

File: test/test-main.js [hunk 36]
```diff
@@ -0,0 +1,10 @@
+global.DATABASE_URL = 'mongodb://localhost/basic-auth-test'
+global.PORT = '8081'
```
Comment (benjaminEwhite) [inline test/test-main.js]: Let's use [this approach](https://github.com/Thinkful-Ed/node-restaurants-app-mongoose/blob/feature/with-tests/server.js#L133) for DATABASE_URL and PORT in runserver so we can avoid globals.

Comment (benjaminEwhite) [reply test/test-main.js]: FYI/FWIW, that change in the restaurants app didn't make it's way into basic-auth app yet, so there was no way for you to know.

File: test/test-main.js [hunk 37]
```diff
@@ -0,0 +1,10 @@
...
+
+var runServer = require('../server').runServer;
+
+before(function(done) {
```
Comment (benjaminEwhite) [inline test/test-main.js]: ditto re: done

File: test/test-users.js [hunk 38]
```diff
@@ -0,0 +1,402 @@
...
+          res.type.should.equal('application/json');
+          res.charset.should.equal('utf-8');
+          res.body.should.be.an('array');
+          res.body.length.should.equal(0);
+        });
+    });
+  });
+});
+
+describe('Test /user endpoints with prepopulated users', () => {
```
Comment (benjaminEwhite) [inline test/test-users.js]: why test the idea of prepopulated users?

File: test/test-users.js (hunk 12 above)
Comment (benjaminEwhite) [reply test/test-users.js]: See pattern [here](https://github.com/Thinkful-Ed/node-restaurants-app-mongoose/blob/feature/with-tests/test/test-restaurants-integration.js#L82)

File: test/test-users.js [hunk 39]
```diff
@@ -0,0 +1,402 @@
...
+    describe('GET', () => {
+      it('should return a list of users', () => {
+        return chai.request(app).get('/users')
+          .then(res => {
+            res.should.have.status(200);
+            res.type.should.equal('application/json');
+            res.charset.should.equal('utf-8');
+
+            res.body.should.be.an('array');
+            res.body.length.should.equal(2); //there are 2 dummy users in our DB
```
Comment (benjaminEwhite) [inline test/test-users.js]: When would this assertion ever fail? Seems like we could easily get rid of this one.

File: test/test-users.js [hunk 40]
```diff
@@ -0,0 +1,402 @@
...
+          lastName: 'user',
+        };
+        return chai.request(app).post('/users').send(newUser)
+          .then(res => {
+            // Check that an empty object is returned
+            res.should.have.status(201);
+            res.type.should.equal('application/json');
+            res.charset.should.equal('utf-8');
+
+            res.should.have.header('location');
```
Comment (benjaminEwhite) [inline test/test-users.js:109]: is that passport related? mind adding a comment in code?

File: test/test-users.js [hunk 41]
```diff
@@ -0,0 +1,402 @@
...
+          .catch(err => {
+            // If the request fails, make sure it contains the error
+            let res = err.response;
+            res.should.have.status(409);
+            res.type.should.equal('application/json');
+            res.charset.should.equal('utf-8');
+            res.body.should.be.an('object');
+
+            res.body.should.have.property('name');
+            res.body.name.should.equal('MongoError');
```
Comment (benjaminEwhite) [inline test/test-users.js]: Cross-reference comments about exposing `MongoError`s.

File: test/test-users.js [hunk 42]
```diff
@@ -0,0 +1,402 @@
...
+          return chai.request(app).get('/users/me').auth(johnDoe.username, 'bad-password')
+            .then(spy)
+            .catch(function(err) {
+              err.response.should.have.status(401);
+            })
+            .then(function() {
+              spy.should.not.have.been.called();
+            });
+        });
+        it('successful authentication should return a single user', () => {
```
Comment (benjaminEwhite) [inline test/test-users.js]: prefer to have this come first, just so that conceptually, we're always testing normal case first, then a handful of edge cases second.

Comment (benjaminEwhite) [inline test/test-users.js]: `it('should return an object representing the user if user authenticates')`

File: test/test-users.js [hunk 43]
```diff
@@ -0,0 +1,402 @@
...
+
+      });
+
+      describe('PUT', function() {
+        it('should allow editing a user (john doe) firstName and lastName', function() {
+          let updated = {
+              firstName: 'test john',
+              lastName: 'test doe',
+            }
+          return chai.request(app).put('/users/me').send(updated).auth(johnDoe.username, johnDoe.password)
```
Comment (benjaminEwhite) [inline test/test-users.js]: break this chain up over several lines for readability.

File: test/test-users.js [hunk 44]
```diff
@@ -0,0 +1,402 @@
...
+              // Check DB that the user has been updated
+              result.firstName.should.equal(updated.firstName);
+              result.lastName.should.equal(updated.lastName);
+            });
+        });
+        it('should allow changing a user username', function() {
+            let updated = {
+              username: 'new.email@example.com'
+            }
+            return chai.request(app).put('/users/me').send(updated).auth(johnDoe.username, johnDoe.password)
```
Comment (benjaminEwhite) [inline test/test-users.js]: break this up

File: test/test-users.js [hunk 45]
```diff
@@ -0,0 +1,402 @@
...
+              result.username.should.equal(updated.username);
+            });
+        });
+        it('should not allow changing a user username to an empty string', function() {
+          let spy = chai.spy();
+
+          let updated = {
+            username: ''
+          }
+          return chai.request(app).put('/users/me').send(updated).auth(johnDoe.username, johnDoe.password)
```
Comment (benjaminEwhite) [inline test/test-users.js]: ditto

File: test/test-users.js [hunk 46]
```diff
@@ -0,0 +1,402 @@
...
+            .then(function() {
+              spy.should.not.have.been.called();
+            });
+        });
+        //Note there are hundreds of validation tests that can be run on email validation regex
+        it('should not allow changing a user username to an invalid email', function() {
+          let spy = chai.spy();
+          let updated = {
+            username: 'test.user@example' // no TLD (.com, .org, .gov etc...)
+          }
```
Comment (benjaminEwhite) [inline test/test-users.js]: colon

File: test/test-users.js [hunk 47]
```diff
@@ -0,0 +1,402 @@
...
+          return chai.request(app).put('/users/me').send(updated).auth(johnDoe.username, johnDoe.password)
```
Comment (benjaminEwhite) [inline test/test-users.js]: break this up

File: test/test-users.js [hunk 48]
```diff
@@ -0,0 +1,402 @@
...
+              spy.should.not.have.been.called();
+            });
+        });
+        it('should not allow changing a user username existing username', function() {
+          //login as john doe and attempt to update username to jane doe
+          let spy = chai.spy();
+          let updated = {
+            username: janeDoe.username
+          }
+          return chai.request(app).put('/users/me').send(updated).auth(johnDoe.username, johnDoe.password)
```
Comment (benjaminEwhite) [inline test/test-users.js]: ditto

File: test/test-users.js [hunk 49]
```diff
@@ -0,0 +1,402 @@
...
+              err.response.should.have.status(422); // 422 Unprocessable Entity
+            })
+            .then(function() {
+              spy.should.not.have.been.called();
+            });
+        });
+      });
+
+      describe('DELETE', function() {
+        it('successful authentication should return a single user', () => {
```
Comment (benjaminEwhite) [inline test/test-users.js]: description is wrong, no?

File: test/test-users.js [hunk 50]
```diff
@@ -0,0 +1,402 @@
...
+          return chai.request(app).delete('/users/me').auth(johnDoe.username, johnDoe.password)
```
Comment (benjaminEwhite) [inline test/test-users.js]: break this up

**Review by benjaminEwhite — COMMENTED** [review]
This looks good overall. A bunch of comments throughout, but main theme is just aligning  style/strategy for tests elsewhere in course.

That means, amongst other things, using same strategy for `before/after` routines (returning promises, instead of `done`) etc.
*

File: server.js (hunk 1 above)
Comment (cklanac) [inline server.js]: Accidentally inherited from the old template when I was testing stuff out... I'll remove

File: test/test-users.js (hunk 3 above)
Comment (cklanac) [reply test/test-users.js]: Cool. Makes sense. will update.

File: test/test-users.js (hunk 2 above)
Comment (cklanac) [reply test/test-users.js]: Well spotted

File: test/test-users.js (hunk 15 above)
Comment (cklanac) [reply test/test-users.js]: It was a test we had in the SUP app, so I recreated it here. I'd love to drop it since it makes the tests more complex than necessary

File: test/test-users.js (hunk 5 above)
Comment (cklanac) [reply test/test-users.js]: Ah, much better approach than mine. Will fix

File: test/test-users.js (hunk 4 above)
Comment (cklanac) [reply test/test-users.js]: No objection to `mongoose.connection.dropDatabase();`. Do you prefer the dropDatabase and ensureIndexes over the and User.remove({})??

File: test/test-users.js (hunk 6 above)
Comment (cklanac) [reply test/test-users.js]: Will fix

File: test/test-users.js (hunk 7 above)
Comment (cklanac) [reply test/test-users.js]: ditto with fix

File: test/test-users.js (hunk 14 above)
Comment (cklanac) [reply test/test-users.js]: Originally part of a larger describe block. But we're dropping this test now anyway

File: test/test-users.js (hunk 8 above)
Comment (cklanac) [reply test/test-users.js]: Dropping this test

File: test/test-users.js (hunk 38 above)
Comment (cklanac) [reply test/test-users.js]: We pre-populate the test before each run. When I broke out the test for an empty database, I updated this describe block to help set it apart  from the other. Now that we're dropping the empty test we can remove the "prepopulated" wording. But we will still drop the database before each test so we'll need to prepopulate with dummy users.

File: test/test-users.js (hunk 9 above)
Comment (cklanac) [reply test/test-users.js]: gotcha

File: test/test-users.js (hunk 10 above)
Comment (cklanac) [reply test/test-users.js]: ditto

File: test/test-users.js (hunk 11 above)
Comment (cklanac) [reply test/test-users.js]: gotcha

File: test/test-users.js (hunk 12 above)
Comment (cklanac) [reply test/test-users.js]: Dropping the database also wipes the indexes, this is another legit approach that is supposed to make the tests faster. But we can go with my "approach #1" which drops and ensures the indexes are recreated.

File: test/test-users.js (hunk 46 above)
Comment (cklanac) [reply test/test-users.js]: Not sure what you mean?

File: users/router.js (hunk 27 above)
Comment (cklanac) [reply users/router.js:48]: Could catch and re-throw?

File: users/router.js (hunk 28 above)
Comment (cklanac) [reply users/router.js]: It caused an error when the username is not valid but cause `callback(null, false, {message: 'Incorrect username'});` returns an empty string to the .then(). Might be another way to fix it but I'll need to look into it

File: test/test-users.js (hunk 15 above)
Comment (benjaminEwhite) [reply test/test-users.js]: Sweet let's drop it!

File: test/test-users.js (hunk 4 above)
Comment (benjaminEwhite) [reply test/test-users.js]: Into with having an `ensureIndexes` method on `User`. Maybe call it `ensureTestIndexes` or something like that just so it's clear method is for testing.

Would like to retain the before, beforeEach, afterEach, after steps mentioned above:

```
before(function() {
    return runServer(TEST_DATABASE_URL);
  });

  beforeEach(function() {
    return promiseThatSeedsDb();
  });

  afterEach(function() {
    return tearDownDb();
  });

  after(function() {
    return closeServer();
  })

```

So the db gets seeded before each test (`ensureIndex`) and tear it down after each test.

Comment (benjaminEwhite) [reply test/test-users.js]: Wait, i misunderstood ensureIndex. Hold on.

Comment (benjaminEwhite) [reply test/test-users.js]: Yeah, so if I understand the `ensureIndexes` thing right, then I think we'd want to do:

```js
beforeEach(function() {
  // ensure indexes
  // seed database with test data
})
```

File: test/test-users.js (hunk 12 above)
Comment (benjaminEwhite) [reply test/test-users.js]: Ah okay, so in that case, we don't really need an `afterEach` routine. Yeah, that makes sense. Maybe not so important in this current app cause just one model, but would this evolve into this sorta thing:

```js
Promise.all([User.remove({}).exec(), OtherModel.remove({}).exec(), YetAnotherModel.remove({}).exec()])
  .then(() => seedDatabase());
```

File: test/test-users.js (hunk 46 above)
Comment (benjaminEwhite) [reply test/test-users.js]: Hah. Semi-colon is what i meant.

File: users/router.js (hunk 27 above)
Comment (benjaminEwhite) [reply users/router.js:48]: Think that's probably better.

File: users/models.js (hunk 26 above)
Comment (cklanac) [reply users/models.js]: Anytime the user is created or saved the `UserSchema.pre('save',...)` runs. If the password has been changed, then it hash it, then save the new user to the DB. This eliminates the need to add bcrypt to both POST and PUT as well as the the test page.

File: server.js (hunk 1 above)
Comment (cklanac) [inline server.js]: done

File: test/test-users.js (hunk 3 above)
Comment (cklanac) [inline test/test-users.js]: done

File: test/test-users.js (hunk 2 above)
Comment (cklanac) [inline test/test-users.js]: done

File: test/test-users.js (hunk 15 above)
Comment (cklanac) [inline test/test-users.js]: done

File: test/test-users.js (hunk 12 above)
Comment (cklanac) [reply test/test-users.js]: Decided to stick with dropping the DB on afterEach and then running ensureIndex

File: test/test-users.js (hunk 39 above)
Comment (cklanac) [reply test/test-users.js]: It is currently the only test of GET /users/. I suppose as we grow the app, we'll be able to replace it... but I'd keep it for now

File: test/test-users.js (hunk 40 above)
Comment (cklanac) [reply test/test-users.js:109]: Commented and added a representation to the body.

File: test/test-users.js (hunk 21 above)
Comment (cklanac) [reply test/test-users.js:134]: We're verifying the spy wasn't called. I post request without a username should return an error so the `.then(spy)` shouldn't be called
